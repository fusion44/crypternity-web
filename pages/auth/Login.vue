<template lang="pug">
    div
      v-form(v-model="valid" ref="form" lazy-validation)
        h1 Login!
        v-text-field(v-model="username" label="Username" required
                     v-bind:error-messages="errorMessage" :rules="[rules.username]")
        v-text-field(v-model="password" label="Password" type="password" required
                     v-bind:error-messages="errorMessage" :rules="[rules.password]")
        v-btn(@click="login" :disabled="!valid") Login
        br 
        router-link(to="/auth/register" color="grey") Register a new account
</template>

<script>
import { mapGetters } from "vuex"

export default {
  data() {
    return {
      username: "",
      password: "",
      valid: false,
      error: false,
      errorMessage: [],
      rules: {
        username: value => {
          if (value.length < 3) return "3 characters minimum"
          return true
        },
        password: value => {
          if (value.length < 6) return "6 characters minimum"
          return true
        }
      }
    }
  },
  computed: mapGetters(["loginState"]),
  methods: {
    login() {
      if (this.$refs.form.validate()) {
        // Reset any error messages
        this.$data.errorMessage = []

        this.$store
          .dispatch("login", {
            username: this.$data.username,
            password: this.$data.password
          })
          .then(result => {
            if (result.error) {
              this.$data.errorMessage = ["Bad credentials"]
            } else {
              this.$data.errorMessage = []
            }
          })
      }
    }
  }
}
</script>

<style>

</style>
