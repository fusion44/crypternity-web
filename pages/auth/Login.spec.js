import toBeType from "jest-tobetype"
import { mount } from "vue-test-utils"
import Login from "./Login"

expect.extend(toBeType)

describe("Login", () => {
  let wrapper

  beforeEach(() => {
    wrapper = mount(Login)
  })

  it("renders the correct markup", () => {
    let html = wrapper.html()
    expect(html).toMatchSnapshot()
  })

  it("validates username length", () => {
    // Returns either true or a string with error message
    let res = wrapper.vm.rules.username("12")
    expect(res).not.toBeFalsy()
    expect(res).toBeType("string")
    expect(res).toContain("3")

    res = wrapper.vm.rules.username("1234")
    expect(res).toBeTruthy()
    expect(res).toBe(true)
  })

  it("validates password length", () => {
    let res = wrapper.vm.rules.password("12345")
    expect(res).not.toBeFalsy()
    expect(res).toBeType("string")
    expect(res).toContain("6")

    res = wrapper.vm.rules.password("123456")
    expect(res).toBeTruthy()
    expect(res).toBe(true)
  })
})
