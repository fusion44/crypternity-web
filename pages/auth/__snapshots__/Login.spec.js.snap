// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login renders the correct markup 1`] = `
<div>
  <v-form lazy-validation="lazy-validation">
    <h1>Login!</h1>
    <v-text-field label="Username" required="required" error-messages="" rules="function username(value) {
          if (value.length < 3) return &quot;3 characters minimum&quot;;
          return true;
        }"></v-text-field>
    <v-text-field label="Password" type="password" required="required" error-messages="" rules="function password(value) {
          if (value.length < 6) return &quot;6 characters minimum&quot;;
          return true;
        }"></v-text-field>
    <v-btn disabled="true">Login</v-btn>
    <br>
    <router-link to="/auth/register" color="grey">Register a new account</router-link>
  </v-form>
</div>
`;
